lightning/src/ln/onion_utils.rs

```diff
@@ -848,49 +843,95 @@ mod tests {
+
+		// Note that the resulting onion differs from the test vectors due to our
+		// `construct_onion_packet_with_init_noise` implementation.
+		assert_eq!(packet.encode(), hex::decode("0002EEC7245D6B7D2CCB30380BFBE2A3648CD7A942653F5AA340EDCEA1F283686619F7F3416A5AA36DC7EEB3EC6D421E9615471AB84C75A1B96D6BDE2F3BC5450EFB8B10E7EE7D90585AAF084E13A66E3B2BD0E926CCD8D263072206DBED0234F6505E21E282ABD8C0E4F5B9FF8042800BBAB065036EADD0149B37F27DDE664725A49866E052E809D2B0198AB9610FAA656BBF4EC516763A59F8F42C171B179166BA38958D4F51B3E73E6C7DD77BA8F874C673E50D1E7C243185154C2FC9B3EEB2F6D1EE1FA4F9D21A163EDF0FB445D5C8D7D688D6DD9C541762BF922EA0F2E798894B55B46EB286551DEFFD0D7506DFE7842FFA1E3C8A34449305546275AFB233ACF4323974680779F1A964CFE65EC142A1495A86666F439AD1D11EA1BB52A2B349A71C778F911ED23D39080D6B9D207A503EE6F6299C92D6361BC335FCBF9B5CD44747AADCE2CE6069CFDC3D671DAEF9F8AE590CF93D957C9E873E9A1BC62D9640DC8FC39C14902D49A1C80239B6C5B7FD91D05878CBF5FFC7DB2569F47C43D6C0D27C438ABFF276E87364DEB8858A37E5A62C446AF95D8B786EAF0B5FCF78D98B41496794F8DCAAC4EEF34B2ACFB94C7E8C32A9E9866A8FA0B6F2A06F00A1CCDE569F97EEC05C803BA7500ACC96691D8898D73D8E6A47B8F43C3D5DE74458D20EDA61474C426359677001FBD75A74D7D5DB6CB4FEB83122F133206203E4E2D293F838BF8C8B3A29ACB321315100B87E80E0EDB272EE80FDA944E3FB6084ED4D7F7C7D21C69D9DA43D31A90B70693F9B0CC3EAC74C11AB8FF655905688916CFA4EF0BD04135F2E50B7C689A21D04E8E981E74C6058188B9B1F9DFC3EEC6838E98809535C2244B757343CFC5E2AAD8CBF22AA2C4A268591974BD98206A5B508BE5E3CA8F956BDC346A51898153F18173F1C77DACFF1FD78B790E253D3C998AA22404331F340AF9567A8A398A1155B6A2754B8C7817BE793A96FB04A9F63D9963B82ECA10272B1357D85340715B0FDAE9C68D89FAAD20547C82432B1F6F7584F3DFA39F6DC83FE17351C1D15F7D8A13B64DA9CC2C7CCE74A5AAAE9C58FEB9947258416DFF072C34F7ECCA923701CD59CEC5F6B8AEFBCBED564EF78C84277A1F8C09AED9A1764EFB8BF6FCE557C4030BD65DF7576B00A8F22ABA138E9441F7EB133724E9A95AC6581E1E48992714FC1DB3C34B05BB7AF5BC2E1DEABC069F1E70B5E60943E2F4EC01D8B5F2DEE087879832EEC4D5060FA08B81B786DB3B4B10693CE3368442C8CAF17FB156182967986EF4171E089069142B8332AD26B67BFA48DD1752B594A408D686151AE0C8A1B12E5D76493E2A6A5C7411716AB10D1075F28941C6E0FD68282A1F0737273D8E704C9975CF99B8D2453B88C2C6DF29019FFDD80688EDCABF4C4D41F24F6790D7065A303AD171483184854A699886D610289D2D9FE68D56E09ACCAD90FF3D65783420662FF3F312F9E2B7E8F5C425E3B37D675556BD710EA2ABDDF24F6C9D291235F622827E0BE4843258BA014B45C2D520A277AB9EFCA564D3CA25D50BF8958CA51C0FBEB9B61C0454E582005505E1A63AEBDB9F0C1E63F1EB39EBCC69F57E9A3030B47FD9754DDA90EBF701288FC982418D5CD36049BC166EC3E808CEA2895243EFD9A7657C58CB33C93DADC75C06E211201EBE43B8BDA0F4BFF5B39DE232DAA33D6FEF9A1125149067B00C2DF4D53549CD6330A2307C4ADD4F11127A677F1C9ED24A7B9A48848A561BB05001D5736994975E5324CB02E46169D9598E5EF442DF21509746B50D1B971A4E53190BEDC858F7DFA6D656016F0D026FF49A874327329356551D504EBBC9C47A441D427247DF9896177C020F3CFC5D6570A292E3D492BF494F557B29EE6A92BB4EAE7E5B3BAFABD9845E984130F0327FB38A8067BAFADF5744FAA446C16").unwrap());
```

Comment:
The result from packet.encode() differs from the test vectors, despite the `payloads`, `onion_keys` and `associated_data` are the equal to the test vector values.

I figured that this must be due to our `construct_onion_packet_with_init_noise` implementation, as our old implementation of the test vectors packet also differed from those vectors' onion. Is this correct?

I haven't had time to go through `construct_onion_packet_with_init_noise` deeply enough to understand why this is the case, so if anyone knows why, I'd be super thankful if you have time to explain why. Else I'll take the time to dig deeper into this later :)

Reply:
Yeah agree, i'll debug it further and let you know the result. I'm a bit packed with work this week due to internal deadlines, but will try to squeeze it in as soon as I have time. Thanks for offering to look into it, but no worries, I'll do it as soon as I have time :)